### DISCUSSION TITLE
Serverpod 2.0.1: notInSet() not working - flutter: Converting object to an encodable object failed: _Set len:10

### DISCUSSION THREAD

**Discussion by KelvinTK** [serverpod/serverpod#2574] [Q&A] [answered]
Environment Info:
VS Code, Flutter, Serverpod 2.0.1, on Windows 11
Debugging on a windows VM

Full Error:
flutter: Failed call: visits.updateVisits
flutter: Converting object to an encodable object failed: _Set len:10

Error info:
Trying to exclude some table IDs using serverpods .notInSet(), server-side code is being generated without issues but the calls to the server are failing with the conversion error mentioned above.

Code highlight:
Future<List<Visit>?> updateVisits(Session session, int personnelId,  Set<int> visIds) async {
    int cnt = await Visit.db.count( session,  where: (v) =>v.personnelId.equals(personnelId) );
    if (cnt > visIds.length) {
      return await Visit.db.find(session,  where: (v) => v.id.notInSet(visIds) & v.personnelId.equals(personnelId) );
   }
return null;
}

**Comment by KelvinTK** [comment]
Sorry, typo on Code highlight:
the method is a nullable Future list of type Visit

**Comment by SandPod** [comment]
Hey @KelvinTK 👋 

If I understand your example correctly, the error appear on the flutter side of the code.

Could you include the code for where you call the server method in your example?
Do you see any errors in the Server logs?

**Comment by KelvinTK** [comment]
Hi @SandPod 

Thank you for responding. Sorry I have already changed my code but from the flutter side, I was calling it (ansychronously) like this:

var newVis = await client.visitManager.updateVisits(user.id!, visitIds);
//visitIds is a set of integers

The flutter exception was smth like 'UnsupportedJson...' if I remember well.

Please note - inSet() is also not working - tested both with Serverpod 2.0.1 and Serverpod 2.0.2

**Reply by KelvinTK** [reply]
... and there are no errors in the server logs

**Comment by SandPod** [comment]
Since there are no error logs on the server suspect the error is in the flutter app, possibly the serialization of Set.

How is `visitIds` defined? What happens if you hard code a set of integers passed to that method instead of passing in the variable? For example 

```dart 
var mySet = {1, 2};
var newVis = await client.visitManager.updateVisits(user.id!, mySet);
```

Do you still get the error in that case?

**Reply by KelvinTK** [reply]
Thats not working either. I have just created a test method to see and here's the error:
Exception has occurred.
JsonUnsupportedObjectError (Converting object to an encodable object failed: _Set len:2)

Casting in the server method is also not working

However, hard-coding the set like ...notInSet({1, 2}); directly (i.e. not making the set a method argument) like this:

` Future<List<Visit>?> test(Session session) async {
    // Set<int> idss = ids.toSet();
    return await Visit.db.find(
      session,
      where: (v) => v.id.notInSet({1, 2}),
    );
  }`

is working!

The set, however, will not have the same ids all the time:(

**Comment by KelvinTK** [comment, marked as answer]
Finally found a solution:

`Future<List<Visit>?> test(Session session, List<int> ids) async { 
    Set<int> idss = {};
    idss.addAll(ids);
    return await Visit.db.find(
      session,
      where: (v) => v.id.notInSet(idss),
    );
  }
`

Thanks for the help @SandPod ...had almost given up. 

Other better solutions are still welcome))

**Reply by Isakdl** [reply]
FYI The `Set` type is not supported by Serverpod as a parameter in endpoints or as a type in models.

E.g. this is not valid:
`Future<List?> updateVisits(Session session, int personnelId, Set visIds)`

**Reply by KelvinTK** [reply]
I finally realized that), hence the list and then to set in the solution that I posted

**Reply by SandPod** [reply]
Aaah good hint @Isakdl and good find @KelvinTK!

I see now that we have two issues related to this: https://github.com/serverpod/serverpod/issues/1915, https://github.com/serverpod/serverpod/issues/1784
